File: Platform/Workspace/CSharp/Allors.Workspace.Blazor/Forms/RoleField.cs
namespace Allors.Workspace.Blazor
{
    using System;
    using Meta;
    using Microsoft.AspNetCore.Components;
    using Microsoft.AspNetCore.Components.Forms;

    public partial class RoleField : Field, IDisposable
    {
        private ISessionObject @object;
        private string name;
        private bool disable;
        private bool required;
        private string label;

        [Parameter]
        public ISessionObject Object
        {
            get => this.@object ?? (ISessionObject)this.EditContext?.Model;
            set => this.@object = value;
        }

        public bool ExistObject => this.Object != null;

        [Parameter]
        public IRoleType RoleType { get; set; }

        public bool CanRead => this.ExistObject && this.Object.CanRead(this.RoleType);

        public bool CanWrite => this.ExistObject && this.Object.CanWrite(this.RoleType);

        public virtual object Model
        {
            get => this.ExistObject ? this.Object.Get(this.RoleType) : null;

            set
            {
                if (this.ExistObject)
                {

                    if (value is string stringValue && this.EmptyStringIsNull && string.IsNullOrWhiteSpace(stringValue))
                    {
                        value = null;
                    }

                    this.Object.Set(this.RoleType, value);
                }
            }
        }

        [Parameter]
        public string Name
        {
            get => !string.IsNullOrWhiteSpace(this.name)
                    ? this.name
                    : this.RoleType.Name + '_' + this.Object?.Id;
            set => this.name = value;
        }

        [Parameter]
        public string Label
        {
            get => !string.IsNullOrWhiteSpace(this.label) ? this.label : this.RoleType.Name; // TODO: Humanize(this.RoleType.Name);
            set => this.label = value;
        }

        [Parameter]
        public bool ShowLabel { get; set; } = true;

        [Parameter]
        public bool Disable
        {
            get => this.disable || !this.CanWrite;
            set => this.disable = value;
        }

        [Parameter]
        public bool Required
        {
            get => this.required ? this.required : this.RoleType.IsRequired;
            set => this.required = value;
        }

        [Parameter]
        public bool Readonly { get; set; }

        [Parameter]
        public string Hint { get; set; }

        [Parameter]
        public bool Focus { get; set; }

        [Parameter]
        public bool EmptyStringIsNull { get; set; } = true;

        public string TextType
        {
            get
            {
                if (this.RoleType.ObjectType.Id == UnitIds.Integer ||
                    this.RoleType.ObjectType.Id == UnitIds.Decimal ||
                    this.RoleType.ObjectType.Id == UnitIds.Float)
                {
                    return "number";
                }

                return "text";

            }
        }

        public string DataAllorsId => this.Object?.Id.ToString("D");

        public string DataAllorsRoleType => this.RoleType?.Id.ToString("D");

        public FieldIdentifier FieldIdentifier => new FieldIdentifier(this, "Model");

        public void Add(ISessionObject value)
        {
            if (this.ExistObject)
            {
                this.Object.Add(this.RoleType, value);
            }
        }

        public void Remove(ISessionObject value)
        {
            if (this.ExistObject)
            {
                this.Object.Remove(this.RoleType, value);
            }
        }

        void IDisposable.Dispose() => this.Validation.Remove(this);

        public override void Validate(ValidationMessageStore messages)
        {
            if (this.RoleType.IsRequired && this.Model == null)
            {
                messages.Add(this.FieldIdentifier, $"{this.RoleType.Name} is required");
            }
            else
            {
                messages.Clear(this.FieldIdentifier);
            }
        }

        protected override void OnAfterRender(bool firstRender)
        {
            if (firstRender)
            {
                this.Validation?.Add(this);
            }
        }
    }
}

File: dotnet/core/workspace/typescript/libs/angular/core/src/export/forms/rolefield.ts
// tslint:disable: directive-selector
// tslint:disable: directive-class-suffix
import { AfterViewInit, Component, Input, OnDestroy, QueryList, ViewChildren, Directive } from '@angular/core';
import { NgForm, NgModel } from '@angular/forms';

import { RoleType, assert, humanize } from '@allors/meta/system';
import { ISessionObject } from '@allors/domain/system';

import { Field } from './Field';

@Directive()
export abstract class RoleField extends Field implements AfterViewInit, OnDestroy {
  @Input()
  public object: ISessionObject;

  @Input()
  public roleType: RoleType;

  @Input()
  assignedRoleType: RoleType;

  @Input()
  derivedInitialRole: ISessionObject;

  // tslint:disable-next-line:no-input-rename
  @Input('name')
  public assignedName: string;

  // tslint:disable-next-line:no-input-rename
  @Input('disabled')
  public assignedDisabled: boolean;

  // tslint:disable-next-line:no-input-rename
  @Input('required')
  public assignedRequired: boolean;

  // tslint:disable-next-line:no-input-rename
  @Input('label')
  public assignedLabel: string;

  @Input()
  public readonly: boolean;

  @Input()
  public hint: string;

  @Input()
  public focus: boolean;

  @Input()
  public emptyStringIsNull = true;

  @ViewChildren(NgModel) private controls: QueryList<NgModel>;

  private id = 0;

  constructor(private parentForm: NgForm) {
    super();
    // TODO: wrap around
    this.id = ++Field.counter;
  }

  get ExistObject(): boolean {
    return !!this.object;
  }

  get model(): any {
    if (this.ExistObject) {
      if (this.assignedRoleType) {
        if (this.object.hasChangedRole(this.assignedRoleType)) {
          return this.object.get(this.assignedRoleType);
        }

        if (this.object.isNew && this.derivedInitialRole) {
          return this.derivedInitialRole;
        }
      }

      return this.object.get(this.roleType);
    }

    return undefined;
  }

  set model(value: any) {
    if (this.ExistObject) {
      if (this.emptyStringIsNull && value === '') {
        value = null;
      }

      if (value && value.toISOString) {
        if (typeof value.toISOString === 'function') {
          value = (value as any).toISOString();
        }
      }

      if (this.roleType.objectType.isDecimal) {
        value = (value as string)?.replace(',', '.');
      }

      if (this.assignedRoleType) {
        if (this.object.isNew && this.derivedInitialRole && this.derivedInitialRole === value) {
          this.object.set(this.assignedRoleType, null);
        } else {
          this.object.set(this.assignedRoleType, value);
        }
      } else {
        this.object.set(this.roleType, value);
      }
    }
  }

  get canRead(): boolean | undefined {
    return this.object?.canRead(this.roleType);
  }

  get canWrite(): boolean | undefined {
    return this.object?.canWrite(this.assignedRoleType ?? this.roleType);
  }

  get textType(): string {
    if (this.roleType.objectType.name === 'Integer' || this.roleType.objectType.name === 'Float') {
      return 'number';
    }

    return 'text';
  }

  get name(): string {
    return this.assignedName ? this.assignedName : this.roleType.name + '_' + this.id;
  }

  get label(): string {
    return this.assignedLabel ? this.assignedLabel : humanize(this.roleType.name);
  }

  get required(): boolean {
    if (this.assignedRequired) {
      return this.assignedRequired;
    }

    if (this.object) {
      return this.roleType.isRequired(this.object.objectType);
    }

    return false;
  }

  get disabled(): boolean {
    return !this.canWrite || !!this.assignedDisabled;
  }

  get canRestore(): boolean {
    return this.ExistObject && this.assignedRoleType && this.object.hasChangedRole(this.assignedRoleType);
  }

  restore(): void {
    this.ExistObject && this.object?.restoreRole(this.assignedRoleType);
  }

  public add(value: ISessionObject) {
    if (this.ExistObject) {
      this.object.add(this.roleType, value);
    }
  }

  public remove(value: ISessionObject) {
    if (this.ExistObject) {
      this.object.remove(this.roleType, value);
    }
  }

  public ngAfterViewInit(): void {
    if (!!this.parentForm) {
      this.controls.forEach((control: NgModel) => {
        this.parentForm.addControl(control);
      });
    }
  }

  public ngOnDestroy(): void {
    if (!!this.parentForm) {
      this.controls.forEach((control: NgModel) => {
        this.parentForm.removeControl(control);
      });
    }
  }

  get dataAllorsId(): string {
    return this.object?.id;
  }

  get dataAllorsRoleType(): string {
    return this.roleType.id;
  }
}
